DB/Messaging integration improved

LeonArcher opened the pull request.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/MessageController.java:
```diff
@@ -26,39 +27,60 @@
+        TaskRunner<Map<String, Date>> initTaskRunner =
+                new TaskRunner<>(Executors.newSingleThreadExecutor());
+
+        initTaskRunner.runTask(
```
Какая то сложная схема, лучше в getNewMessages стартовать эту задачу если надо и после её выполнения уже смотреть что по сообщениям.

LeonArcher replied:
Т.к. возвращаемый результат отличается от основного (заполненная мапа вместо списка сообщений), необходим новый, отдельный TaskRunner. Вообще говоря, непонятно, как запуск задачи в getNewMessages поможет обойти эту проблему.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/MessageController.java:
```diff
@@ -26,39 +27,60 @@
+    Map<String, Date> skipToDateTable = null;
 
-    public MessageController(Handler uiHandler, NetworkObjectLayer network) {
+    public MessageController(Handler uiHandler, NetworkObjectLayer network, DBHandler db) {
```
На класс @MainThread лучше повесить.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -323,9 +328,11 @@ public synchronized Message getMessage(int id) {
+    // Getting All Messages by receiverId
     @Nullable
-    public synchronized List<Message> getAllMessagesByReceiverId(int receiverId) {
+    public synchronized List<Message> getAllMessagesByReceiverId(int receiverId)
```
Т.к. возможно исключение надо оборачивать закрывание курсора в try-finally

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -390,9 +425,9 @@ public synchronized Message getMessage(int id) {
+    // Getting All Messages
     @Nullable
-    public synchronized List<Message> getAllMessages() {
+    public synchronized List<Message> getAllMessages() throws ParseException {
```
Лучше наверно кидать какой нибудь SQLDataException

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -17,12 +16,16 @@
 
 
-@WorkerThread
+// @WorkerThread
```
Можно повесить аннотацию на каждый отдельный метод

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/ReceiveMessagesTask.java:
```diff
@@ -3,56 +3,102 @@
-    private final int skipToId;
+    private final String targetId;
+
+    private final Map<String, Date> skipToDateTable;
```
Лучше не шарить мапу между разными классами а держать её в том месте где стартуются таски на получение сообщений. Обновлять мапу надо в callback.
В таску подавать информацию необходимую для фильтрации.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/MessageController.java:
```diff
@@ -26,48 +29,89 @@
+                    public void onSuccess(List<Message> result) {
+                        int sz = result.size();
+                        if (sz > 0) {
+                            skipToDateTable.put(targetId, result.get(sz - 1).getDate());
```
Лучше фильтрацию skipToDate унести из самой ReceiveMessageTask в MessageController. Тогда не надо будет расшаривать мапу skipToDate.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/MessageController.java:
```diff
@@ -26,48 +29,89 @@
-        Integer val = skipToIdTable.get(key);
-        int skipToId = (val != null) ? val : 0;
+        // if no skipToDateTable defined yet
+        if (skipToDateTable == null) {
```
Тут поидее надо класть таску в очередь на выполнение, чтобы она выполнилась когда инициализация пройдет.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/LastMessageDateProvider.java:
```diff
@@ -0,0 +1,79 @@
+
+        // try to init table from database, init with empty if failed
+        try {
+            newSkipToDateTable = new ConcurrentHashMap<>();
```
Тут не нужен ConcurrentHashMap, у тебя работа с мапой и так защищена

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/LastMessageDateProvider.java:
```diff
@@ -0,0 +1,79 @@
+            }
+
+        } catch (SQLDataException ex) {
+            newSkipToDateTable = new ConcurrentHashMap<>();
```
Можно в начале метода её создавать. Исключение молча нехорошо проглатывать. Надо писать хотя-бы в лог.

LeonArcher replied:
А если в процессе заполнения мапы возникнет исключение? Тогда лучше пустую мапу возвращать, чем полу-заполненную.

cypt replied:
Исключение всё равно нало залогировать

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/LastMessageDateProvider.java:
```diff
@@ -0,0 +1,79 @@
+            newSkipToDateTable = new ConcurrentHashMap<>();
+        }
+
+        skipToDateTable = newSkipToDateTable;
```
Еще одно лишнее присваивание

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/LastMessageDateProvider.java:
```diff
@@ -0,0 +1,79 @@
+        return skipToDateTable.get(targetId);
+    }
+
+    // updates table only if more recent date passed
```
Javadoc можно сделать через 
/**
*/

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/messaging/MessageController.java:
```diff
@@ -26,48 +25,61 @@
 
-    public MessageController(Handler uiHandler, NetworkObjectLayer network) {
+    // provide date to skip incoming messages up to
+    LastMessageDateProvider lastMessageDateProvider;
```
private, final?

cypt reviewed (commented):
lgtm

cypt commented:
lgtm

K-Mike commented:
lgtm

LeonArcher committed:
Database handler class integration improved. Skipping already existing in
DB messages while receiving (using Date field). Additional minor changes,
improvements and bug fixes.

cypt commented:
lgtm

K-Mike commented:
lgtm

LeonArcher commented:
lgtm

cypt commented:
lgtm

LeonArcher commented:
lgtm

cypt commented:
lgtm

K-Mike commented:
lgtm

cypt commented:
lgtm

cypt commented:
approvals/lgtm

LeonArcher commented:
lgtm

LeonArcher commented:
approvals/lgtm
